lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUTransformCharFilter.java

```diff
@@ -0,0 +1,322 @@
+
+  ICUTransformCharFilter(Reader in, Transliterator transform) {
+    this(in, transform, DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY);
+  }
+
+  /**
+   * Construct new {@link ICUTransformCharFilter} with the specified {@link Transliterator}, backed by
+   * the specified {@link Reader}.
+   * @param in input source
+   * @param transform used to perform transliteration
+   * @param maxRollbackBufferCapacityHint used to control the maximum size to which this
+   * {@link ICUTransformCharFilter} will buffer and rollback partial transliteration of input sequences.
+   * The provided hint will be converted to an enforced limit of "the greatest power of 2 (excluding '1')
+   * less than or equal to the specified value". Specifying a negative value allows the rollback buffer to
+   * grow indefinitely (equivalent to specifying {@link Integer#MAX_VALUE}). Specifying "0" (or "1", in practice)
+   * disables rollback. Larger values can in some cases yield more accurate transliteration, at the cost of
+   * performance and resolution/accuracy of offset correction.
+   * This is intended primarily as a failsafe, with a relatively large default value of {@value ICUTransformCharFilter#DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY}.
+   * See comments "To understand the need for rollback" in private method:
+   * {@link Transliterator#filteredTransliterate(com.ibm.icu.text.Replaceable, Position, boolean, boolean)}
+   */
+  ICUTransformCharFilter(Reader in, Transliterator transform, int maxRollbackBufferCapacityHint) {
+    super(in);
+    this.transform = ICUTransformFilter.optimizeForCommonCase(transform);
+    if (maxRollbackBufferCapacityHint < 0) {
+      this.maxRollbackBufferCapacity = Integer.MAX_VALUE;
+      this.rollbackBuffer = new char[DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY];
+    } else {
+      // greatest power of 2 (excluding "1") less than or equal to the specified hint
+      this.maxRollbackBufferCapacity = Integer.highestOneBit(maxRollbackBufferCapacityHint - 1) << 1;
+      if (this.maxRollbackBufferCapacity == 0) {
+        this.rollbackBuffer = null;
+      } else {
+        this.rollbackBuffer = new char[DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY];
+      }
+    }
+  }
+
+  /**
+   * Reads characters into a portion of an array. This method will block until some input is available, an I/O error
+   * occurs, or the end of the stream is reached.
+   *
+   * @param cbuf
+   *          Destination buffer
+   * @param off
+   *          Offset at which to start storing characters
+   * @param len
+   *          Maximum number of characters to read
+   * @return The number of characters read, or -1 if the end of the stream has been reached
+   * @throws IOException
+   *           If an I/O error occurs
+   */
+  @Override
+  public int read(char[] cbuf, int off, int len) throws IOException {
+    if (off < 0) throw new IllegalArgumentException("off < 0");
+    if (off >= cbuf.length) throw new IllegalArgumentException("off >= cbuf.length");
```

Comment:
This error message probably won't be meaningful to most callers - how about something like "offset exceeeds buffer size"?

Reply:
+1, pushed change that also changes exception type to `IndexOutOfBoundsException`, as prescribed in `Reader` class abstract method javadocs. (I had copied bounds-checking code from [ICUNormalizer2CharFilter.java](https://github.com/apache/lucene-solr/blob/master/lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUNormalizer2CharFilter.java#L70))